Feature/publication state

PhilippeFerreiraDeSousa opened the pull request:
Draft : Seul les membres du club ont accès aux brouillons dans la liste des publications du club.
Scheduled : Cette publication apparaîtra seulement sur le calendrier uPont.
Published : Cette publication sera publique sur uPont, vous pourrez envoyer un mail plus tard.
Emailed: La publication sera publiée et envoyée par mail à tous les utilisateurs de uPont qui suivent le club.

PhilippeFerreiraDeSousa committed:
Add states Draft, Scheduled, Published to events only first to test

PhilippeFerreiraDeSousa committed:
Déplacement de l'état de publication sur Post. Arrangement formulaire de création d'event et news

PhilippeFerreiraDeSousa committed:
Creation and modification of events work, cannot modify to inferior state

PhilippeFerreiraDeSousa committed:
Publications with ribbon, draft events from other club members don't appear

PhilippeFerreiraDeSousa committed:
La modification des events/news se fait directement à l'endroit des publications

PhilippeFerreiraDeSousa committed:
add cancelation of publication edit

PhilippeFerreiraDeSousa committed:
remove alertify ask for mail

PhilippeFerreiraDeSousa committed:
Add ClubUser fixture, remove deprecated doctrine assertion constraint

PhilippeFerreiraDeSousa committed:
Change publication ticket color to publication state color

PhilippeFerreiraDeSousa committed:
Fix multiple pagination, add /events/check-dates/{slug} to check for overlapping events at the creation or modification of an event. Check for authorization on drafts on /events and /newsitems, create all personnel messages as 'Published'

PhilippeFerreiraDeSousa committed:
Small front factorization

PhilippeFerreiraDeSousa committed:
Back propre sur les requêtes d'events/newsitems qui ne sont pas des brouillons d'autrui

PhilippeFerreiraDeSousa committed:
Adaptation aux users de l'administration

PhilippeFerreiraDeSousa committed:
Oops typos

LouisTrezzini commented on back/app/DoctrineMigrations/Version20171015200834.php:
```diff
@@ -0,0 +1,34 @@
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE Post ADD publicationState VARCHAR(255) DEFAULT \'Published\', DROP send_mail');
```
Est-ce qu'on a pas envie d'attribuer le bon `publishedState` suivant la valeur actuelle de `send_mail` lors de la migration ?

PhilippeFerreiraDeSousa replied:
Oui oui, faut faire des boucles php ici avec des addSql partout ? o:)

LouisTrezzini commented on back/src/KI/CoreBundle/Controller/ResourceController.php:
```diff
@@ -12,12 +12,12 @@ class ResourceController extends LikeableController
      * @return Response
      */
-    public function getAll($auth = false)
+    public function getAll($auth = false, array $findBy = [])
```
Pourquoi un $findBy sur le getAll des resources ?

PhilippeFerreiraDeSousa replied:
Je l'utilise pour requêter les messages à /newsitems?name=message
```
if ($request->query->get('name') == 'message') {
      $findBy = array('name' => 'message');
      return $this->getAll($this->is('EXTERIEUR'), $findBy);
}
```
En soi donner "message" comme titre de news aux messages de l'accueil est très douteux

LouisTrezzini commented on back/src/KI/CoreBundle/Helper/PaginateHelper.php:
```diff
@@ -33,11 +33,6 @@ public function paginateData(EntityRepository $repository, array $findBy = [])
 
-        // On s'assure de bien recevoir des arrays
-        foreach ($findBy as $key => $value) {
-            $findBy[$key] = array($value);
```
À garder avec un truc du genre `$aRecords = is_array($aRecords) ? $aRecords : array($aRecords);`

LouisTrezzini commented on back/src/KI/PublicationBundle/Controller/EventsController.php:
```diff
@@ -63,6 +69,9 @@ public function getEventsAction()
     {
         $event = $this->getOne($slug);
+        if ($event->getPublicationState() == 'Draft' && !$this->isClubMember($event->getAuthorClub())) {
+            return $this->json('Tu n\'es pas autorisé à lire ce brouillon !', 403);
```
`throw $this->createAccessDeniedException('You cannot access this page!');`

LouisTrezzini commented on back/src/KI/PublicationBundle/Controller/EventsController.php:
```diff
@@ -39,9 +39,15 @@ public function setContainer(ContainerInterface $container = null)
+            $this->getUser()->getId(),
+            $request->query->get('publicationState'),
+            $request->query->get('limit'),
+            $request->query->get('page'));
```
Problème d'indentation

PhilippeFerreiraDeSousa replied:
Regarde bien, les lignes doublements indentées sont en argument de `getAllowedEvents`

LouisTrezzini commented on back/src/KI/PublicationBundle/Controller/EventsController.php:
```diff
@@ -63,6 +69,9 @@ public function getEventsAction()
     public function getEventAction($slug)
     {
         $event = $this->getOne($slug);
+        if ($event->getPublicationState() == 'Draft' && !$this->isClubMember($event->getAuthorClub())) {
```
Le `publicationState` devrait être slug si possible -> `draft`

Du coup, si c'est `scheduled`, tout le monde peut accéder à la fiche de l'événement ? Quelle différence avec `published` ?

PhilippeFerreiraDeSousa replied:
Oui, je vais les mettre en slug. C'était comme ça à la base par cohérence avec l'entryMethod des Events 'Shotgun', 'Libre', 'Ferie' mais c'est plus très justifié maintenant

Ce qui est published est sur le newsfeed et le calendrier, ce qui est scheduled n'est que sur le newsfeed des membre de l'asso qui peuvent l'éditer et sur le calendrier (mais accessible par recherche dans la topbar et par requête au back)

LouisTrezzini commented on back/src/KI/PublicationBundle/Controller/EventsController.php:
```diff
@@ -161,6 +170,36 @@ public function deleteEventAction($slug)
+        $endDate = $request->query->get('endDate');
+        $events = $this->repository->findBy(array('publicationState' => array('Scheduled', 'Published', 'Emailed')));
+        $matchedEvents = array();
+        foreach ($events as $event) {
```
qui a dit que c'était une bonne idée de boucler sur l'intégralité des entrées de la base de données ?

PhilippeFerreiraDeSousa replied:
Oui, il faut mettre cette logique dans une requête DQL dans le repository

LouisTrezzini commented on back/src/KI/PublicationBundle/Controller/EventsController.php:
```diff
@@ -161,6 +170,36 @@ public function deleteEventAction($slug)
+    {
+        $startDate = $request->query->get('startDate');
+        $endDate = $request->query->get('endDate');
+        $events = $this->repository->findBy(array('publicationState' => array('Scheduled', 'Published', 'Emailed')));
```
la notation `array` est dépréciée, utilise `[]` et indente

LouisTrezzini commented on back/src/KI/PublicationBundle/Controller/NewsitemsController.php:
```diff
@@ -32,9 +34,21 @@ public function setContainer(ContainerInterface $container = null)
      * @Method("GET")
      */
-    public function getNewsitemsAction()
+    public function getNewsitemsAction(Request $request)
```
Est-ce que tous les newsitems ne seraient pas des événements ? Pourquoi s'emmerder à dupliquer le code partout ?

PhilippeFerreiraDeSousa replied:
Ca aurait été trop logique, Newsitem et Event héritent tous deux de Post

LouisTrezzini commented on back/src/KI/PublicationBundle/Controller/NewsitemsController.php:
```diff
@@ -32,9 +34,21 @@ public function setContainer(ContainerInterface $container = null)
+    public function getNewsitemsAction(Request $request)
     {
-        return $this->getAll($this->is('EXTERIEUR'));
+        if ($request->query->get('name') == 'message') {
```
Est-ce que les messages sont vraiment utiles sur uPont ?

PhilippeFerreiraDeSousa replied:
Non effectivement, on aurait pu en faire quelque chose mais ils n'ont même pas d'existence réel dans la bdd. Mais bon ils ne font de mal à personne où ils sont

LouisTrezzini commented on back/src/KI/PublicationBundle/Controller/NewsitemsController.php:
```diff
@@ -56,6 +70,9 @@ public function getNewsitemsAction()
     public function getNewsitemAction($slug)
     {
         $newsitem = $this->getOne($slug, $this->is('EXTERIEUR'));
+        if ($newsitem->getPublicationState() == 'Draft' && !$this->isClubMember($newsitem->getAuthorClub())) {
```
idem

LouisTrezzini commented on back/src/KI/PublicationBundle/Controller/NewsitemsController.php:
```diff
@@ -56,6 +70,9 @@ public function getNewsitemsAction()
     {
         $newsitem = $this->getOne($slug, $this->is('EXTERIEUR'));
+        if ($newsitem->getPublicationState() == 'Draft' && !$this->isClubMember($newsitem->getAuthorClub())) {
+            return $this->json('Tu n\'es pas autorisé à lire ce brouillon !', 403);
```
idem

LouisTrezzini commented on back/src/KI/PublicationBundle/DataFixtures/ORM/LoadEvent.php:
```diff
@@ -18,10 +18,10 @@ public function load(ObjectManager $manager)
         $event->setAuthorClub($this->getReference('club-ki'));
         $event->setAuthorUser($this->getReference('user-trancara'));
         $event->setEntryMethod('Libre');
+        $event->setPublicationState('Emailed');
```
slugify

LouisTrezzini commented on back/src/KI/PublicationBundle/DataFixtures/ORM/LoadEvent.php:
```diff
@@ -35,10 +35,10 @@ public function load(ObjectManager $manager)
         $event->setAuthorClub($this->getReference('club-ki'));
         $event->setAuthorUser($this->getReference('user-muzardt'));
         $event->setEntryMethod('Libre');
+        $event->setPublicationState('Scheduled');
```
slugify

LouisTrezzini commented on back/src/KI/PublicationBundle/DataFixtures/ORM/LoadEvent.php:
```diff
@@ -51,11 +51,11 @@ public function load(ObjectManager $manager)
         $event->setAuthorClub($this->getReference('club-mediatek'));
         $event->setAuthorUser($this->getReference('user-trancara'));
         $event->setEntryMethod('Shotgun');
+        $event->setPublicationState('Emailed');
```
slugify

LouisTrezzini commented on back/src/KI/PublicationBundle/DataFixtures/ORM/LoadEvent.php:
```diff
@@ -68,10 +68,10 @@ public function load(ObjectManager $manager)
         $event->setAuthorClub($this->getReference('club-pep'));
         $event->setAuthorUser($this->getReference('user-guerinh'));
         $event->setEntryMethod('Libre');
+        $event->setPublicationState('Draft');
```
slugify

LouisTrezzini commented on back/src/KI/PublicationBundle/DataFixtures/ORM/LoadEvent.php:
```diff
@@ -83,10 +83,10 @@ public function load(ObjectManager $manager)
         $event->setAuthorClub($this->getReference('club-bde'));
         $event->setAuthorUser($this->getReference('user-dziris'));
         $event->setEntryMethod('Libre');
+        $event->setPublicationState('Published');
```
slugify

LouisTrezzini commented on back/src/KI/PublicationBundle/DataFixtures/ORM/LoadEvent.php:
```diff
@@ -102,11 +102,11 @@ public function load(ObjectManager $manager)
         $event->setAuthorClub($this->getReference('club-bda'));
         $event->setAuthorUser($this->getReference('user-donat-bb'));
         $event->setEntryMethod('Shotgun');
+        $event->setPublicationState('Scheduled');
```
slugify

LouisTrezzini commented on back/src/KI/PublicationBundle/DataFixtures/ORM/LoadEvent.php:
```diff
@@ -115,11 +115,11 @@ public function load(ObjectManager $manager)
         $event->setAuthorClub($this->getReference('club-ki'));
         $event->setAuthorUser($this->getReference('user-trancara'));
         $event->setEntryMethod('Ferie');
+        $event->setPublicationState('Draft');
```
slugify

LouisTrezzini commented on back/src/KI/PublicationBundle/DataFixtures/ORM/LoadNewsitem.php:
```diff
@@ -17,6 +17,7 @@ public function load(ObjectManager $manager)
         $newsitem->setDate(1414242424);
         $newsitem->setAuthorClub($this->getReference('club-ki'));
         $newsitem->setAuthorUser($this->getReference('user-taquet-c'));
+        $newsitem->setPublicationState('Published');
```
slugify

LouisTrezzini commented on back/src/KI/PublicationBundle/DataFixtures/ORM/LoadNewsitem.php:
```diff
@@ -65,6 +70,20 @@ public function load(ObjectManager $manager)
+        $manager->persist($newsitem);
+
+        $newsitem = new Newsitem();
+        $newsitem->setName('Le béton c\'est bon.');
```
Utilise des guillemets double au lieu de tout échapper...

LouisTrezzini commented on back/src/KI/PublicationBundle/Entity/Post.php:
```diff
@@ -53,12 +53,13 @@ public function getAutoSetUser() { return $this->autoSetUser; }
-     * La publication envoie-t-elle un mail ?
-     * @ORM\Column(name="send_mail", type="boolean", nullable=false)
+     * Etat de la publication [Draft|Scheduled|Published|Emailed]
+     * @ORM\Column(name="publicationState", type="string", nullable=true, options={"default" = "Published"})
```
nullable=true ?

LouisTrezzini replied:
est-ce qu'il y a des colonnes en camelCase dans la BDD ? je pense que c'est plutôt du snake_case

PhilippeFerreiraDeSousa replied:
Je sais pas comment remplir la colonne lors de la migration pour les entrées existentes. Mais ça devrait être nullable=false.

Les variables Doctrine sont en camelCase, les correspondances SQL et les réponses jsonified du back sont en snake_case

LouisTrezzini commented on back/src/KI/PublicationBundle/Entity/Post.php:
```diff
@@ -53,12 +53,13 @@ public function getAutoSetUser() { return $this->autoSetUser; }
      */
-    protected $sendMail = false;
+    protected $publicationState;
+    const STATE_ORDER = array('Draft' => 1, 'Scheduled' => 2, 'Published' => 3, 'Emailed' => 4);
```
short array notation + slug

LouisTrezzini commented on back/src/KI/PublicationBundle/Entity/Post.php:
```diff
@@ -241,15 +242,43 @@ public function upload()
+    {
+        if (array_key_exists($publicationState, $this::STATE_ORDER)) {
+            if (isset($this->publicationState) && $this::STATE_ORDER[$publicationState] < $this::STATE_ORDER[$this->publicationState]) {
+                throw new BadRequestHttpException('Revenir à un état de publication antérieure est interdit !');
```
BadRequestHttpException dans une entité ? ça n'a aucun sens

LouisTrezzini commented on back/src/KI/PublicationBundle/Entity/Post.php:
```diff
@@ -241,15 +242,43 @@ public function upload()
+            return $this->publicationState = $publicationState;
+        }
+        else {
+            throw new BadRequestHttpException('L\'état de la publication doit être Draft, Scheduled, Published ou Emailed !');
```
idem

LouisTrezzini commented on back/src/KI/PublicationBundle/Entity/Post.php:
```diff
@@ -241,15 +242,43 @@ public function upload()
      * @return Post
      */
-    public function setSendMail($sendMail)
+    public function setSendMail()
```
Utilisé quelque part ?

PhilippeFerreiraDeSousa replied:
Nop

LouisTrezzini commented on back/src/KI/PublicationBundle/Entity/Post.php:
```diff
@@ -261,6 +290,6 @@ public function setSendMail($sendMail)
      */
     public function getSendMail()
```
Utilisé quelque part ?

LouisTrezzini commented on back/src/KI/PublicationBundle/Entity/Post.php:
```diff
@@ -241,15 +242,43 @@ public function upload()
-     * @param boolean $sendMail
+     * @return Event
+     */
+    public function setPublicationState($publicationState)
```
Est-ce qu'on veut vraiment pouvoir set le state directement ? Est-ce que ça ne serait pas plus safe d'avoir des méthodes qui font avancer de 1 ?

PhilippeFerreiraDeSousa replied:
Oui, mais du coup on ne peut plus passer par le Form pour modifier l'objet, faut écrire des routes qui font ça et dans le front (post.js) dire d'augmenter de la différence entre l'order de l'état initial et le nouveau. En fait les gens sont pas obilgés de passer par les quatre états et personne le fera, typiquement il vont peut-être le mettre en planifié et le passer à emailed directement. Du coup faire des routes invrémentales et moins justifiées

LouisTrezzini commented on back/src/KI/PublicationBundle/Form/EventType.php:
```diff
@@ -19,11 +19,11 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ->add('startDate')
             ->add('endDate')
             ->add('entryMethod')
+            ->add('publicationState')
```
est-ce qu'on veut vraiment une dropdown pour choisir l'état ?
je verrais plus un bouton qui change au fur et à mesure 

[ Enregistrer mon brouillon | V ]  <-- bouton initial, si on clique sur le chevron vers le bas, il s'étend
|   Programmer l'événement    |  <-- que si c'est un événement
|                   Publier                 |
|           Envoyer un email        |

PhilippeFerreiraDeSousa replied:
Actuellement c'est "Publication" par défaut, "Plannification" n'apparait que pour les events et quand on modifie la news/events les états antérieures sont disabled donc c'est pas loin et plus simple à coder. Faire un chevron qui s'étend ça à l'air cool, tu l'étends ou tu le déplaces vers le bas ?

LouisTrezzini commented on back/src/KI/UserBundle/DataFixtures/ORM/LoadClubUser.php:
```diff
@@ -13,6 +13,13 @@ class LoadClubUserFixture extends AbstractFixture implements OrderedFixtureInter
+        $clubUser->setClub($this->getReference('club-bde'));
+        $clubUser->setUser($this->getReference('user-matthias.dreveton'));
+        $clubUser->setPriority($clubUser->getUser()->getId());
+        $clubUser->setRole('Sex\'Gen');
```
guillemets doubles

LouisTrezzini commented on back/src/KI/UserBundle/Controller/OwnController.php:
```diff
@@ -454,13 +455,14 @@ private function getCourseitems($user = null)
      * @Method("GET")
      */
-    public function getNewsItemsAction()
+    public function getNewsItemsAction(Request $request)
```
étrange

LouisTrezzini commented on back/src/KI/UserBundle/Controller/OwnController.php:
```diff
@@ -365,14 +365,16 @@ protected function getFollowedClubs($user = null)
-            $events = $userRepository->findFollowedEvents($this->getUser()->getId());
+            $events = $this->repository->findFollowedEvents(
+                $this->getUser()->getId(),
+                $request->query->get('publicationState')
```
Pourquoi la requête permet-elle de choisir les states qu'on affiche ?

PhilippeFerreiraDeSousa replied:
Le calendrier ne requête pas les brouillon, si on ne peut requêter que tout après il faut faire une boucle pour enlever le brouillon, j'ai cru comprendre que tu trouvais les boucles php un peu salle quand on peut faire des requêtes dql

LouisTrezzini commented on front/app/js/controllers/users/calendar.js:
```diff
@@ -54,7 +54,7 @@ angular.module('upont')
                 resolve: {
                     events: ['$resource', function($resource) {
-                        return $resource(apiPrefix + 'own/events').query().$promise;
+                        return $resource(apiPrefix + 'own/events?publicationState=Scheduled,Published,Emailed').query().$promise;
```
why

PhilippeFerreiraDeSousa replied:
Les brouillons sont privés, pas sur le calendrier

LouisTrezzini commented on front/app/js/controllers/users/publications/list.html:
```diff
@@ -1,15 +1,17 @@
             ng-repeat="publication in events.data | concat: newsItems.data | orderBy: 'date' | reverse"
-            ng-if="publication.entry_method != 'Ferie'">
+            ng-if="publication.entry_method != 'Ferie' && ($root.hasClub(publication.author_club.slug) || ['Published', 'Emailed'].includes(publication.publication_state))">
+            <up-ribbon ng-repeat="(state, item) in pub_info" class="Ribbon--{{ pub_info[publication.publication_state].color }}" ng-hide="editSlug == publication.slug"
```
pourquoi tu boucles alors que tu sais que `publication.publication_state == state` ?

PhilippeFerreiraDeSousa replied:
Il y a 4 ribbons différents, un membre voit les publications de son club pour les 4 états.
On est dans list.html à l'intérieur d'un `ng-repeat="publication in events.data | concat: newsItems.data | orderBy: 'date' | reverse"` donc publication.publication_state change dans ce feed de publications

LouisTrezzini commented on front/app/js/controllers/users/publications/list.html:
```diff
@@ -1,15 +1,17 @@
+        <article col class="Ticket Ticket--{{ pub_info[publication.publication_state].color }}"
             ng-repeat="publication in events.data | concat: newsItems.data | orderBy: 'date' | reverse"
-            ng-if="publication.entry_method != 'Ferie'">
+            ng-if="publication.entry_method != 'Ferie' && ($root.hasClub(publication.author_club.slug) || ['Published', 'Emailed'].includes(publication.publication_state))">
```
pourquoi tu ne renvoies pas que les bonnes publications aux bons users ? pourquoi il y a de la logique dans le front ?

PhilippeFerreiraDeSousa replied:
On veut juste pas afficher dans le feed les événements planifiés des autres assos mais y laisser accès, vu qu'on utlise la même route ownClub pour le calendrier et qu'on peut l'utiliser pour autre chose plus tard.
On peut simplifier en `$root.hasClub(publication.author_club.slug) || publication.publication_state != 'scheduled'`

LouisTrezzini commented on front/app/js/controllers/users/publications/list.html:
```diff
@@ -116,9 +117,18 @@
             <div up-likes show="false" objet="publication"
-                 url="(publication.start_date ? 'events/' : 'newsitems/') + publication.slug"></div>
+                 url="(publication.start_date ? 'events/' : 'newsitems/') + publication.slug"
+                 ng-if="!['Draft', 'Scheduled'].includes(publication.publication_state)"></div>
```
berk

PhilippeFerreiraDeSousa replied:
Pourquoi ?

LouisTrezzini commented on front/app/js/controllers/users/publications/post.html:
```diff
@@ -29,14 +29,27 @@
+                <select class="Input" ng-model="post.entry_method">
+                    <option value="Libre" selected ng-disabled="modify && initialEntryMethod == 'Shotgun'">Entrée libre</option>
+                    <option value="Shotgun">Shotgun</option>
+                    <option value="Ferie" ng-if="$root.isAdmin" ng-disabled="modify && initialEntryMethod == 'Shotgun'">Ferie</option>
```
ng-if="$root.isAdmin"

...
je pense que tu peux supprimer tout référence à "Ferie" du coup

PhilippeFerreiraDeSousa replied:
C'est pour faire apparaître les vacances scolaires et jours fériés sur le calendrier. Ca aurait plus de sens d'être sur le panneau d'administration mais pour l'instant ça fait de mal à personne ici. On pourra statuer là-dessus plus tard je pense

LouisTrezzini commented on front/app/js/controllers/users/publications/post.html:
```diff
@@ -82,9 +95,10 @@
                         <i class="fa fa-file-o"></i>
                         <span ng-repeat="file in postFiles"> {{ file.name ? file.name : 'Fichier' }} </span>
-                        <button class="InputButton__fileinput" ngf-select="selectFiles($files)" multiple></button>
+                        <button class="InputButton__fileinput" ng-select="selectFiles($files)" multiple></button>
```
ngf-select pour les fichiers...

LouisTrezzini reviewed (commented):
Je n'ai pas eu le courage de tout relire en détail ni de tester en local

PhilippeFerreiraDeSousa reviewed (commented):
Je vais revoir tout ça

PhilippeFerreiraDeSousa commented:
Editer la publucation pour changer l'état dans la dropdown c'était la solution rapide. Je vais réfléchir au système incrémental qui est plus élaboré mais faut que je trouve du temps

PhilippeFerreiraDeSousa committed:
change style

PhilippeFerreiraDeSousa committed:
migrate data send_mail into publicationState, move getEventCheckDatesAction selection logic into repo

PhilippeFerreiraDeSousa committed:
fix unregistered repos, bfix bug changing a publication slug twice in a row and request to checkDates after changing slug

To do : test data migration, correct get to own pubilcations, unpredictable redactor. The calendar library should be changed in the future. Some fonts are missing for some months, event should inherit from newsitem and why the fuck are personnel messages newsitems with title 'message'

PhilippeFerreiraDeSousa committed:
fix data migration down

PhilippeFerreiraDeSousa committed:
Not working version on /club/slug/newsitems but checking totalpages in pagination
